crates/sui-bytecode-src-verifier/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    collections::{BTreeMap, HashMap},
+    fmt::{Debug, Display},
+    path::Path,
+    str::FromStr,
+};
+
+use move_core_types::account_address::AccountAddress;
+use move_package::{compilation::compiled_package::CompiledPackage, BuildConfig};
+use move_symbol_pool::Symbol;
+
+use sui_sdk::{
+    rpc_types::{SuiRawData, SuiRawMoveObject, SuiRawMovePackage},
+    ReadApi,
+};
+use sui_types::{
+    base_types::{ObjectID, ObjectIDParseError},
+    error::SuiError,
+};
+
+#[derive(Clone, Debug)]
+pub struct DependencyVerificationResult {
+    pub verified_dependencies: HashMap<AccountAddress, Dependency>,
+}
+
+#[derive(Debug)]
+pub enum DependencyVerificationError {
+    /// Could not resolve Sui addresses for package dependencies
+    ResolutionGraphNotResolved(anyhow::Error),
+    /// Could not convert a dependencies' resolved Sui address to a Sui object ID
+    ObjectIdFromAddressFailure(ObjectIDParseError),
+    /// Could not read a dependencies' on-chain object
+    DependencyObjectReadFailure(anyhow::Error),
+    /// Dependency object does not exist or was deleted
+    SuiObjectRefFailure(SuiError),
+    /// Dependency address contains a Sui object, not a Move package
+    ObjectFoundWhenPackageExpected(ObjectID, SuiRawMoveObject),
+    /// A local dependency was not found
+    ///
+    /// params:  package, module
+    LocalDependencyNotFound(Symbol, Option<Symbol>),
+    /// Local dependencies have a different number of modules than on-chain
+    ///
+    /// params:  expected count, on-chain count
+    ModuleCountMismatch(usize, usize),
+    /// A local dependency module did not match its on-chain version
+    ///
+    /// params:  package, module, address, expected, found
+    ModuleBytecodeMismatch(String, String, AccountAddress, Vec<u8>, Vec<u8>),
+}
+
+impl Display for DependencyVerificationError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Debug::fmt(&self, f)
+    }
+}
+
+#[derive(Debug)]
+pub struct BytecodeSourceVerifier<'a> {
+    pub verbose: bool,
+    rpc_client: &'a ReadApi,
+}
+
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct Dependency {
+    pub symbol: String,
+    pub module_bytes: BTreeMap<String, Vec<u8>>,
+}
+
+impl<'a> BytecodeSourceVerifier<'a> {
+    pub fn new(rpc_client: &'a ReadApi, verbose: bool) -> Self {
+        BytecodeSourceVerifier {
+            verbose,
+            rpc_client,
+        }
+    }
+
+    /// Verify that all local Move package dependencies' bytecode matches
+    /// the bytecode at the address specified on the Sui network we are publishing to.
+    pub async fn verify_deployed_dependencies(
+        &self,
+        build_config: &BuildConfig,
```

Comment:
I don't think you should need `build_config` or `path` if you already have the `CompiledPackage`--the logic below should be all doable via iterating over `compiled_package.deps_compiled_units`

Reply:
Did this refactor on [move-deps-nopath](https://github.com/MystenLabs/sui/compare/move-deps-nopath) branch.  It does work but i think the resulting code is a bit nastier